src/main/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapper.java

```diff
@@ -212,120 +188,146 @@ protected Explicit<Boolean> ignoreMalformed(BuilderContext context) {
-                            .name(name)
-                            .mappedFieldType(mappedFieldType)
-                            .multiFields(multiFieldsBuilder)
-                            .copyTo(copyToBuilder)
-                            .ignoreMalformed(ignoreMalformed)
-                            .stored(stored.get())
-                            .hasDocValues(hasDocValues.get())
-                            .vectorDataType(vectorDataType.getValue())
-                            .knnMethodContext(knnMethodContext)
-                            .build();
-                    return new LuceneFieldMapper(createLuceneFieldMapperInput);
-                }
 
-                return new MethodFieldMapper(
+                return new ModelFieldMapper(
                     name,
                     mappedFieldType,
                     multiFieldsBuilder,
                     copyToBuilder,
                     ignoreMalformed,
                     stored.get(),
                     hasDocValues.get(),
-                    knnMethodContext
+                    modelDao,
+                    indexCreatedVersion
                 );
             }
 
-            String modelIdAsString = this.modelId.get();
-            if (modelIdAsString != null) {
-                // Because model information is stored in cluster metadata, we are unable to get it here. This is
-                // because to get the cluster metadata, you need access to the cluster state. Because this code is
-                // sometimes used to initialize the cluster state/update cluster state, we cannot get the state here
-                // safely. So, we are unable to validate the model. The model gets validated during ingestion.
+            boolean isResolvedNull = resolvedKNNMethodContext == null;
+            if (isResolvedNull) {
+                resolvedKNNMethodContext = this.knnMethodContext.getValue();
+                setDefaultSpaceType(resolvedKNNMethodContext, vectorDataType.getValue());
+                validateSpaceType(resolvedKNNMethodContext, vectorDataType.getValue());
+                validateDimensions(resolvedKNNMethodContext, vectorDataType.getValue());
+                validateEncoder(resolvedKNNMethodContext, vectorDataType.getValue());
+            }
 
-                return new ModelFieldMapper(
+            // If the field mapper is using the legacy context and being constructed from another field mapper,
+            // the settings will be empt. See https://github.com/opendistro-for-elasticsearch/k-NN/issues/288. In this
+            // case, the input resolvedKNNMethodContext will be null and the settings wont exist (so flat mapper should
+            // be used). Otherwise, we need to check the setting.
+            boolean isSettingPresent = KNNSettings.IS_KNN_INDEX_SETTING.exists(context.indexSettings());
+            if (isResolvedNull && (!isSettingPresent || !KNNSettings.IS_KNN_INDEX_SETTING.get(context.indexSettings()))) {
```

Comment:
Let me try. It is a little tricky to simplify because of upstream issue where merge does not have access to settings (see https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/mapper/ParametrizedFieldMapper.java#L130 and https://github.com/opendistro-for-elasticsearch/k-NN/issues/288)

Reply:
Please take a shot at it. The condition may look easy to read right now, but in next few months it will be hard. Or give an extensive java docs on this so that it is easy to understand.